plugin/pkg/auth/authorizer/rbac/bootstrappolicy/controller_policy.go

```diff
@@ -59,6 +59,16 @@ func buildControllerRoles() ([]rbac.ClusterRole, []rbac.ClusterRoleBinding) {
 	controllerRoleBindings := []rbac.ClusterRoleBinding{}
 
 	addControllerRole(&controllerRoles, &controllerRoleBindings, rbac.ClusterRole{
+		// role for the cloud providers to list kubernetes nodes, required for
+		// GetAllCurrentZones(). See: https://github.com/kubernetes/kubernetes/issues/50115
+		// Remove when node labels are secured so we can use them to
+		// determine cluster vs non-cluster instances
+		ObjectMeta: metav1.ObjectMeta{Name: saRolePrefix + "cloud-provider"},
```

Comment:
Listing nodes isn't particularly sensitive, but I think this is a poorly defined role that is likely to amass unrelated permissions. Is there a reason not to make this and the service account specific to GCE (since the client and usage is specific to GCE)?

What's the plan for moving the GCE PV provisioner to run out of tree, to the point where this is no longer needed as core policy?

Separately, is scalability concerned by fully listing all node API objects on every call to CreateDisk/CreateRegionalDisk/CreateVolume on large clusters? (@wojtek-t , @shyamjvs?)

Reply:
As part of CSI, the GCE volume plugin will be moved out of tree, but it is still a few releases out.

In terms of scalability, we should add some some caching eventually so we don't need to call this every time we provision a disk.  I guess the difference between getting K8s nodes and GCE instances is that the gce api was filtered, so all nodes were not getting returned.  Listing all k8s nodes could cause memory usage to momentarily spike.

Reply:
Sorry I'm lacking context here. FWIU the code, the cloud-provider controller is listing k8s node objects on calls to createDisk. From scalability perspective, imo we should be using informer cache instead of making those list calls (any reason why we're not doing it currently?). Also from @msau42 's comment above, I'm wondering if all k8s nodes need to be listed (or just some with a selector)?

Reply:
List node calls can be time-consuming on large clusters. For e.g on our 5k-node clusters we see a 99%ile latency of ~1.3s (https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-scale-performance/58/artifacts/APIResponsiveness_density_2017-11-03T21:50:19Z.json) which can slow down the createDisk operation (though I don't know how slow other operations are relatively).

Reply:
For context, previously what happened during CreateDisk() is that we would call gce.Instances.List() across all zones in the region, truncated to one page of results each.  But this method returns gce instances that are not managed by Kubernetes.  So the change we want to make here is to change the call to kube.Nodes.List() to only get the nodes that we manage.  It is possible that the the new method will return more nodes to process than the old method since the results are not being truncated.

The cloud provider library currently doesn't have the plumbing in place to pass in informers, but that can be investigated.  I don't believe the difference would show up as a significant performance issue.  Disk creation is already slow, and is not measured in our tests.

Reply:
Makes sense.

> Listing all k8s nodes could cause memory usage to momentarily spike.

Yes, for reasonably large clusters (it's about 20MB for 5k-node cluster). It can be further aggravated if there are multiple parallel CreateDisk calls - though I don't know how much that can happen in practice.

> The cloud provider library currently doesn't have the plumbing in place to pass in informers, but that can be investigated. I don't believe the difference would show up as a significant performance issue. Disk creation is already slow, and is not measured in our tests.

Sounds reasonable. Saw a TODO for moving it to watch later on.

Reply:
I suggested a way to reduce the no. of listed nodes in https://github.com/kubernetes/kubernetes/pull/52322#discussion_r149496088. Besides that, there's one more way by which you can reduce further. Assuming you already know the names of all zones in the region (which sounds like the case from earlier), all you need to do is just list one instance with that label value. You can do this by setting `limit=1` param in your list call. This will totally eliminate the memory spike - but at the cost of multiple (but faster) list calls.